File: words-api/Lib/BridgeTypes/Source.cs
// Project: words-api
// File: Source.cs
// 
// Author: Ethan Kisiel (ethan.a.kisiel@gmail.com)
// 
// File Created: 23 08 2025 23:08:24
// Last Modified: 23 08 2025 23:08:20
// 
// Modified By: Ethan Kisiel (ethan.a.kisiel@gmail.com>)
// 
// Copyright 2025 - 2025 Ethan Kisiel, Ethan Kisiel

namespace words_api.Lib.Enums;

public class Source
{
    public const string Default = "A"; // (Unspecified / reserved)
    public const string Beeson = "B"; // C.H. Beeson, A Primer of Medieval Latin, 1925 (Bee)
    public const string Cassell = "C"; // Charles Beard, Cassell's Latin Dictionary 1892 (CAS)
    public const string Adams = "D"; // J.N. Adams, Latin Sexual Vocabulary, 1982 (Sex)
    public const string Stelten = "E"; // L.F. Stelten, Dictionary of Eccles. Latin, 1995 (Ecc)
    public const string Deferrari = "F"; // Roy J. Deferrari, Dictionary of St. Thomas Aquinas, 1960 (DeF)
    public const string Gildersleeve = "G"; // Gildersleeve + Lodge, Latin Grammar 1895 (G+L)
    public const string Collatinus = "H"; // Collatinus Dictionary by Yves Ouvrard
    public const string Leverett = "I"; // Leverett, F.P., Lexicon of the Latin Language, Boston 1845
    public const string J = "J"; // (Unspecified / reserved)
    public const string Calepinus = "K"; // Calepinus Novus, modern Latin, by Guy Licoppe (Cal)
    public const string Lewis = "L"; // Lewis, C.S., Elementary Latin Dictionary 1891
    public const string Latham = "M"; // Latham, Revised Medieval Word List, 1980
    public const string Nelson = "N"; // Lynn Nelson, Wordlist
    public const string Oxford = "O"; // Oxford Latin Dictionary, 1982 (OLD)
    public const string Souter = "P"; // Souter, A Glossary of Later Latin to 600 A.D., Oxford 1949
    public const string Other = "Q"; // Other, cited or unspecified dictionaries
    public const string PlaterWhite = "R"; // Plater & White, A Grammar of the Vulgate, Oxford 1926
    public const string LewisShort = "S"; // Lewis and Short, A Latin Dictionary, 1879 (L+S)
    public const string Translation = "T"; // Found in a translation -- no dictionary reference
    public const string DuCange = "U"; // Du Cange
    public const string Vademecum = "V"; // Vademecum in opus Saxonis - Franz Blatt (Saxo)
    public const string Guess = "W"; // My personal guess
    public const string Temp = "Y"; // Temp special code
    public const string User = "Z"; // Sent by user -- no dictionary reference

    // X,      --  General or unknown or too common to say
    // A,
    // B,      --  C.H.Beeson, A Primer of Medieval Latin, 1925 (Bee)
    // C,      --  Charles Beard, Cassell's Latin Dictionary 1892 (CAS)
    // D,      --  J.N.Adams, Latin Sexual Vocabulary, 1982 (Sex)
    // E,      --  L.F.Stelten, Dictionary of Eccles. Latin, 1995 (Ecc)
    // F,      --  Roy J. Deferrari, Dictionary of St. Thomas Aquinas, 1960 (DeF)
    // G,      --  Gildersleeve + Lodge, Latin Grammar 1895 (G+L)
    // H,      --  Collatinus Dictionary by Yves Ouvrard
    // I,      --  Leverett, F.P., Lexicon of the Latin Language, Boston 1845
    // J,
    // K,      --  Calepinus Novus, modern Latin, by Guy Licoppe (Cal)
    // L,      --  Lewis, C.S., Elementary Latin Dictionary 1891
    // M,      --  Latham, Revised Medieval Word List, 1980
    // N,      --  Lynn Nelson, Wordlist
    // O,      --  Oxford Latin Dictionary, 1982 (OLD)
    // P,      --  Souter, A Glossary of Later Latin to 600 A.D., Oxford 1949
    // Q,      --  Other, cited or unspecified dictionaries
    // R,      --  Plater & White, A Grammar of the Vulgate, Oxford 1926
    // S,      --  Lewis and Short, A Latin Dictionary, 1879 (L+S)
    // T,      --  Found in a translation  --  no dictionary reference
    // U,      --  Du Cange
    // V,      --  Vademecum in opus Saxonis - Franz Blatt (Saxo)
    // W,      --  My personal guess
    // Y,      --  Temp special code
    // Z       --  Sent by user --  no dictionary reference
}

File: words-frontend/src/lib/models/codes/Source.ts
export type SourceValue =
    'X' | 'A' | 'B' | 'C' | 'D' | 'E' | 'F' | 'G' | 'H' | 'I' | 'J' | 'K' |
    'L' | 'M' | 'N' | 'O' | 'P' | 'Q' | 'R' | 'S' | 'T' | 'U' | 'V' | 'W' |
    'Y' | 'Z';

export class Source {

    private static general = 'X';
    private static unspecified = 'A';
    private static beeson = 'B';
    private static cassell = 'C';
    private static adams = 'D';
    private static stelten = 'E';
    private static deferrari = 'F';
    private static gildersleeve = 'G';
    private static collatinus = 'H';
    private static leverett = 'I';
    private static bracton = 'J';
    private static calepinus = 'K';
    private static lewis = 'L';
    private static latham = 'M';
    private static nelson = 'N';
    private static oxford = 'O';
    private static souter = 'P';
    private static other = 'Q';
    private static platerWhite = 'R';
    private static lewisShort = 'S';
    private static translation = 'T';
    private static u = 'U';
    private static vademecum = 'V';
    private static guess = 'W';
    private static temp = 'Y';
    private static user = 'Z';

    static getLongForm(value: SourceValue, detailed=false): string {
        switch (value) {
            case this.general:
                return 'General or unknown or too common to say';
            case this.unspecified:
                return 'Unspecified';
            case this.beeson:
                return (!detailed ?
                    'Beeson' :
                    'C.H. Beeson, A Primer of Medieval Latin, 1925 (Bee)');
            case this.cassell:
                return (!detailed ?
                    'Cassell' :
                    'Charles Beard, Cassell\'s Latin Dictionary 1892 (Cas)');
            case this.adams:
                return (!detailed ?
                    'Adams' :
                    'J.N. Adams, Latin Sexual Vocabulary, 1982 (Sex)');
            case this.stelten:
                return (!detailed ?
                    'Stelten' :
                    'L.F. Stelten, Dictionary of Ecclesiastical Latin, 1995 (Ecc)');
            case this.deferrari:
                return (!detailed ?
                    'Deferrari' :
                    'Roy J. Deferrari, Dictionary of St. Thomas Aquinas, 1960 (DeF)');
            case this.gildersleeve:
                return (!detailed ?
                    'Gildersleeve' :
                    'Gildersleeve + Lodge, Latin Grammar 1895 (G+L)');
            case this.collatinus:
                return (!detailed ?
                    'Collatinus' :
                    'Collatinus Dictionary by Yves Ouvrard');
            case this.leverett:
                return (!detailed ?
                    'Leverett' :
                    'Leverett, F.P., Lexicon of the Latin Language, Boston 1845');
            case this.bracton:
                return (!detailed ?
                    'Bracton' :
                    'Bracton: De Legibus Et Consuetudinibus Angliae');
            case this.calepinus:
                return (!detailed ?
                    'Calepinus' :
                    'Calepinus Novus, modern Latin, by Guy Licoppe (Cal)');
            case this.lewis:
                return (!detailed ?
                    'Lewis' :
                    'Lewis, C.S., Elementary Latin Dictionary 1891');
            case this.latham:
                return (!detailed ?
                    'Latham' :
                    'Latham, Revised Medieval Word List, 1980 (Latham)');
            case this.nelson:
                return (!detailed ?
                    'Nelson' :
                    'Lynn Nelson, Wordlist (Nel)');
            case this.oxford:
                return (!detailed ?
                    'Oxford' :
                    'Oxford Latin Dictionary, 1982 (OLD)');
            case this.souter:
                return (!detailed ?
                    'Souter' :
                    'Souter, A Glossary of Later Latin to 600 A.D., Oxford 1949');
            case this.other:
                return (!detailed ?
                    'Other' :
                    'Other, cited or unspecified dictionaries');
            case this.platerWhite:
                return (!detailed ?
                    'Plater & White' :
                    'Plater & White, A Grammar of the Vulgate, Oxford 1926 (Plater)');
            case this.lewisShort:
                return (!detailed ?
                    'Lewis & Short' :
                    'Lewis and Short, A Latin Dictionary, 1879 (L+S)');
            case this.translation:
                return (!detailed ?
                    'Translation' :
                    'Found in a translation — no dictionary reference');
            case this.u:
                return 'U (unspecified)';
            case this.vademecum:
                return (!detailed ?
                    'Vademecum' :
                    'Vademecum in opus Saxonis — Franz Blatt (Saxo)');
            case this.guess:
                return (!detailed ?
                    'Guess' :
                    'My personal guess, mostly obvious extrapolation (Whitaker)');
            case this.temp:
                return (!detailed ?
                    'Temp' :
                    'Temp special code');
            case this.user:
                return (!detailed ?
                    'User' :
                    'Sent by user — no dictionary reference, mostly John White of Blitz Latin');
            default:
                return 'General or unknown or too common to say';
        }
    }
}